orchestrator/test_runner/src/erc_721_happy_path.rs

```diff
@@ -0,0 +1,638 @@
+    // randomly and may actually make it impossible for that test to pass
+    // by random re-allocation of powers. If we had 5 or 10 validators
+    // instead of 3 this wouldn't be a problem. But with 3 not even 1% of
+    // power can be reallocated to the down validator before things stop
+    // working. We'll settle for testing that the initial valset (generated
+    // with the first block) is successfully updated
+    if !validator_out {
+        for _ in 0u32..2 {
+            test_valset_update(web30, contact, &mut grpc_client, &keys, gravity_address).await;
+        }
+    } else {
+        wait_for_nonzero_valset(web30, gravity_address).await;
+    }
+
+    // generate an address for coin sending tests, this ensures test imdepotency
+    let user_keys = get_user_key();
+
+    info!("testing erc20 deposit");
+    // the denom and amount of the token bridged from Ethereum -> Cosmos
+    // so the denom is the gravity<hash> token name
+    // Send a token 3 times
+    for _ in 0u32..3 {
+        test_erc20_deposit_panic(
+            web30,
+            contact,
+            &mut grpc_client,
+            user_keys.cosmos_address,
+            gravity_address,
+            erc20_address,
+            100u64.into(),
+            None,
+            None,
+        )
+        .await;
+    }
+
+    let event_nonce = get_event_nonce_safe(gravity_address, web30, *MINER_ADDRESS)
+        .await
+        .unwrap();
+
+    // We are going to submit a duplicate tx with nonce 1
+    // This had better not increase the balance again
+    // this test may have false positives if the timeout is not
+    // long enough. TODO check for an error on the cosmos send response
+    submit_duplicate_erc20_send(
+        event_nonce, // Duplicate the current nonce
+        contact,
+        erc20_address,
+        1u64.into(),
+        user_keys.cosmos_address,
+        &keys,
+    )
+    .await;
+
+    // we test a batch by sending a transaction
+    test_batch(
+        contact,
+        &mut grpc_client,
+        web30,
+        user_keys.eth_address,
+        gravity_address,
+        keys[0].validator_key,
+        user_keys.cosmos_key,
+        erc20_address,
+    )
+    .await;
+}
+
+// Iterates each attestation known by the grpc endpoint by calling `get_attestations()`
+// Executes the input closure `f` against each attestation's decoded claim
+// This is useful for testing that certain attestations exist in the oracle,
+// see the consumers of this function (check_valset_update_attestation(),
+// check_send_to_cosmos_attestation(), etc.) for examples of usage
+//
+// `F` is the type of the closure, a state mutating function which may be called multiple times
+// `F` functions take a single parameter of type `T`, which is some sort of `Message`
+// The type `T` is very important as it dictates how we decode the message
+pub async fn iterate_attestations<F: FnMut(T), T: Message + Default>(
```

Comment:
no need to duplicate this function, import from happy path if you need it.

Reply:
Removed as we are not checking attestations